### ISSUE TITLE
Opt a subtree out of hitTestSelf culling

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#75747]
We often get questions about why hit testing "doesn't work" around transforms. The short answer is that ancestors don't ask their children to hit test if their own hitTestSelf fails (no point asking your children if they were hit if the hit wasn't inside your bounds).

@csells suggested that it would be nice if a subtree could opt-in to not doing that culling. Obviously that would in general be expensive but it would be less expensive than doing a full tree walk (unless someone opted out the entire tree, obviously). I think we would still avoid doing this ever in the Flutter framework since it's better to adjust the widget hierarchy to not have ancestors of the transformed widget be smaller than that widget, but it might be useful in especially elaborate scenes where that is impractical.

One way to do that would be to have HitTestResult have a flag that says whether to apply the hitTestSelf check. Having to check that flag every step of the walk would itself be not cheap, but the cost (which affects every app) might still be worth it?

Another option might be to have a way for a widget to hook into the hit testing of a far ancestor (e.g. via an inherited widget). That would be better in that it would allow the whole thing to be done as a package. Maintaining paint order correctness would be distinctly non-trivial though (maybe even intractable).

Any other ideas?

cc @goderbauer

**Comment by goderbauer** [comment]
A solution for this would also help with the often raised complained that you can't hit-test children drawn in the visual overflow of a stack (see discussion at https://github.com/flutter/flutter/issues/19445).

**Comment by goderbauer** [comment]
> it would be nice if a subtree could opt-in to not doing that culling. Obviously that would in general be expensive but it would be less expensive than doing a full tree walk (unless someone opted out the entire tree, obviously).

I believe choosing the root of the subtree that you want to opt out will not not be easy. The root will not be the transform widget that moves a child out of its bounds. It has to be an ancestor, who has the transformed descendant still in its bounds. Otherwise, the hit test would never reach the transform widget or the transformed child. I am wondering, if in practice that would encourage people to basically opt-out the entire tree...

**Comment by goderbauer** [comment]
Would be interesting to study a use case, that we would enable (or make significantly simpler over e.g. re-organizing the widget tree) by having this mechanism. Do we have one?

**Comment by esDotDev** [comment]
I can certainly provide a number of use cases, just slammed right now. I'll circle back in a cpl weeks when things calm down a bit. 

Generally speaking, it's any time you want to de-couple things, and have some child with children of it's own, that want to hang off the side of that parent.  Think a classic web-menu with pop-out panels that overhang outside a column/row, or an expanding search bar that starts in a column, but then expands downward, or even a panel with a scrollbar, where the scrollbar sits in negative padding. All of these things want to start inside the box-layout of Flutter (column/row/positioned) but then easily expand outside it.

If you can't just offset a child outside yourself, the alternative, usually involves putting the 2 siblings in a stack. In practice, this is not too bad **if you plan from the start for it,** but it's really limiting when requirements change, and you have some existing widget tree, that is not lending itself easily to changing form. 

As a simple example, lets say I have a system for sizing content in a stack, and it looks like this:
![image](https://user-images.githubusercontent.com/736973/108235911-91ce7780-7103-11eb-86ac-3b2caf774285.png)


Immediately we run into this issue. The most maintainable and portable way to build it as an end user of Flutter is:
* Define the content box, in a Stack, at it's true height 300x100
* From within the box, decorate itself by offseting the drag handles slightly into negative space
* Moving the "300x100" into negative space, relative to our center (lets assume this thing needs to be clickable as well).

What we want to write is something really simple like this, where all the sizing info is self-contained within the Box widget:
```
Stack(
children: [
    MyBoxContent(),
    // Corners can sit in negative space, very easy to change size of these btns later.
   // All the sizing logic is here, relative to this box. No co-ordination with parent needed.
    Positoned(
      top: -btnSize/2,  left: -btnSize/2, right: -btnSize/2, bottom: -btnSize/2,
      child: SelectionHandles()
    ),
    /// Bottom SizePanel
    Align(
        alignment: Alignment.bottomCenter, 
        child: FractionalTranslation(
            offset: Offset(0, -1), 
            child: SizePanel()))
  ]
)
```
The parent does need to be concerned with what we're doing in our negative margins, it just sizes the box. The box is free to decorate itself with any "hanging widgets" it needs, without fighting the frameworks hit-detection algorithm.

This creates a super portable 'thing' there is no impact to the Widget tree around, and its super portable. From outside, the code is easily read. So you can imagine how useful this can be when you get some last-minute design request to move some thing slightly outside it's parent bounds... 

If we can't just use the offset trick, we would actually need to make the size of our box something like:
```
// Each box needs this sizing info, from the parent, who really shouldn't care about this cludge...
width: 300 + btnSize/2,
// Add more extra height into the box, just to support hit detection, that the box will need to remove later
height: 300 + btnSize/2 + Insets.paddingLarge + SizePanel.kHeight,
```
And inside the box, we need to size our content in some hack way like:
```
content.width = widget.width - CornersBtns.kBtnSize; // 2 halfs combined, is the 'extra' space passed in from parent
content.height = widget.height -  SizePanel.height; // We need to just know that our parent has passed us this extra space
```
These two layers need to coordinate somehow, either by passing down but sizing info, or defining a bunch of constants. 

This code is extremely representative of what gets written when we are trying to work around these sorts of issues, where anytime something wants to "go outside the box", our only answer is "add more box".

**Comment by Hixie** [comment]
The usual solution to that problem is the Overlay. See e.g. how the text editing toolbar is done.

**Comment by esDotDev** [comment]
Ya, I think that's kinda the problem this would address. It would give us some middle ground. It also shows the problem with giving examples, because the response will tend to be "Use overlay". 

Overlay is quite complicated to learn, and verbose to use. Kinda feels like a sledgehammer when what you more-often need is a chisel? Especially once you add the LayerLink and CompositeTransformFollower into the mix, vs this, which is like 20 characters and a smidge of negative padding and gets all of the following behavior for free. 

Many devs have no clue how to use Overlay, and be immediately scared off of the code. While everyone is familiar with the concept of negative margin, and virtually no one expects btns to stop working when in negative space.

Yes we have to manage sorting, but that's expected from other stacks that work this way (dom, unity, flash), so it's a universal thing almost every developer is somewhat familiar with. Flutter is well-suited to tweaking render order anyways, it can do just-in-time imperative re-orders with dart, or flow the `Flexible` up/down, so it's never much of a problem to put something "on top", it's actually much easier here to control our render order than any of those other stacks because it's all dart code and List<>'s.

Like if you look at it from the perspective of the developer writing code. There is absolutely no attraction to using a more complex approach if the simple one will do. We just want the most options possible, and will always favor the simplest most maintainable KISS approach. If we do have a single thing, that needs a full-screen Modal, we can opt for Overlay. If we have a bunch of multi-select boxes, or some btns we just want to nudge over -10px to line-up visually, Overlay is not particularly well suited. 

A simple Widget that just says "Anything in this sub-tree is clickable, if it's visible" would be a great tool in the toolbox.

**Comment by Hixie** [comment]
Having been on the CSS working group in an earlier life, I have to say I get much much more scared by negative margins than anything like the Overlay. :-)

Anyway I'm not saying Overlay is the answer; if I was, I wouldn't have filed this bug. I'm just saying that that's the way to do it today.

**Comment by esDotDev** [comment]
No worries I get it. Everything is double-edged. Flutter is such an expressive framework though, I always feel it's a shame to be locked too rigidly inside the box (well we're not actually, just our hits are :'( )

By way of example, here is a specific use case I don't think overlay actually supports (negative margins wouldn't either fwiw):
https://user-images.githubusercontent.com/736973/108322563-d6e5be80-7182-11eb-9616-ee4b40db941d.mp4

Here we want the overlay always to have 1px outline, but to otherwise follow the "size" of of the follow-target. 

With overlay, using `CompositeTransformFollower`, the overlay is actually transformed rather than sized to match. In this example the red-box is an OverlayEntry, If using Overlay for the use case above, I would want "Hello" to maintain it's size, and the margins around it change, but instead the whole thing is zoomed. In this case we are just parented inside a `InteractiveViewer`.
https://user-images.githubusercontent.com/736973/108323434-e7e2ff80-7183-11eb-9174-ddd962483279.mp4

This isn't really related to this issue, but a good use case where Overlay should be the perfect solution, but doesn't quite work as needed, but that small gap makes it basically unusable for the use-case.

**Comment by esDotDev** [comment]
A nice thing about having this Widget is that it will let us put some sort of number to this. Curious if we're talking 10ms, 1ms or .1ms when we say it's an expensive thing to apply to the tree.

**Comment by esDotDev** [comment]
As an example of why Overlay is just hard to deal with, I had an idea: "What if I can use a LayerLink, but use just the position and size, but not the scale."... then I could easily implement the Figma Style UI above. Now I have a head-ache after diving into `LeaderLayer`, `ContainerLayer`, `CompositedTransformFollower` and `RenderFollowerLayer`, and still don't feel like I have any clue the answer to my Q. 

I was really just hoping `LayerLink` was a `ChangeNotifier` I could bind to, with Pos, Size, and Scale values, which is really all I need to implement my own in-stack Overlay in a clean and simple way, or do my own positioning within the OverlayEntry, rather than relying on CompositedTransformFollower which is doing things I don't really want.

For now I just have to abandon that approach because the overall system would take weeks of dedicated study to understand deeply.

Another thing to note, even though something like  this should be handled by an overlay-like widget, having it tied to Navigator is problematic. Often, we want to be able to control the sorting order of our Overlays. For example, the Figma "Selection Box: UI will overlay all items within the "Artboard Area" but it will not overlay the panels of the app itself:
![image](https://user-images.githubusercontent.com/736973/108411491-eea25e80-71e5-11eb-8950-bd1d0904aae1.png)

To do this in Flutter with Overlay, you need nested Navigators. This is an invasive and undesired requirement. With a simple LayerLinkDispatcher and a parent Stack, we could achieve the result we need very simply.

**Comment by esDotDev** [comment]
Just to try and gather this all in one place, I believe there's about 158 upvotes on this general issue: 

There is this one, with 113 votes, which is basically asking for this ability (despite the title asking for docs), closed without resolution.
https://github.com/flutter/flutter/issues/19445

This one from 2 yrs ago, with 25 votes, still open
https://github.com/flutter/flutter/issues/31728

This issue with 20 votes, which is making the same request, closed without resolution
https://github.com/flutter/flutter/issues/27587

It just seems to be a repeated cycle of developers saying this is needed to avoid ugly hacks, Flutter team asking for examples, then being un-moved by the examples.

The simple argument this makes our code cleaner, more robust, and easier to debug a lot of the times. It's both a quality of life thing, but also a elegance thing, the alternative is almost always more LOC, more complexity and more time.

**Comment by Vardiak** [comment]
What a delight to see this issue moving forward! I've been asking for this feature for over a year. I think the proposed solution would solve most of our issues regarding hit testing. 😊

**Comment by kartik1225** [comment]
Looking forward for the solution :)

**Comment by esDotDev** [comment]
Here's a use case to get the ball rolling.

This shows 2 issues with flutter layout currently:
* Hit areas can not extend outside their parent
* The size of the child is also clipped

This is a classic, desktop-style dropdown menu, in 65 lines.
https://user-images.githubusercontent.com/736973/111880626-2d0c7400-8972-11eb-9440-adf912a6e8be.mp4

```
// Shows 3 buttons, when rolling a menu should pop up below the button.
class DropDownMenuSpike extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Column(children: [
      Row(
        children: [MenuBtn(), MenuBtn(), MenuBtn()],
      ),
      Expanded(child: Center(child: Text("PAGE CONTENT")))
    ]);
  }
}

class MenuBtn extends StatefulWidget {
  @override
  _MenuBtnState createState() => _MenuBtnState();
}

class _MenuBtnState extends State<MenuBtn> {
  bool _isMouseOver = false;

  @override
  Widget build(BuildContext context) {
    MouseRegion addMouseRegion(Widget child) => MouseRegion(
        child: child,
        onEnter: (_) => setState(() => _isMouseOver = true),
        onExit: (_) => setState(() => _isMouseOver = false));
    Color bgColor = Colors.black.withOpacity(_isMouseOver ? .3 : .1);

    return Container(
      width: 100,
      height: 40,
      color: bgColor,
      child: Stack(
        children: [
          // Main Btn
          addMouseRegion(Center(child: Text("MenuBtn"))),
          // Dropdown Menu
          if (_isMouseOver)
            Transform.translate(
              offset: Offset(0, 30),
              child: addMouseRegion(DropDownMenu()),
            )
        ],
      ),
    );
  }
}

class DropDownMenu extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Container(
      width: 200,
      color: Colors.black.withOpacity(.2),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          TextButton(onPressed: () {}, child: Text("CLICK ME IF YOU CAN!")),
          TextButton(onPressed: () {}, child: Text("CLICK ME IF YOU CAN!")),
          TextButton(onPressed: () {}, child: Text("CLICK ME IF YOU CAN!")),
        ],
      ),
    );
  }
}
```

As written, I don't see any reason why this shouldn't just work. Seems rather clear on it's face: when we mouse over, we show a Column of btns, we offset it by our own height on the Y-axis, dropdown menu is done? And importantly, the code the actually show-hide the menu here is what, 3 or 4 lines? The rest is just content.

This is simple, elegant, portable, composable and easy to maintain. The alternatives are all more cludgy: 
* Make DropDownMenuSpike aware of the dropdown menu positioning 
* Create a wrapper widget (stack) to coordinate between the btns and the popovers 
* Use overlay, 

None of these are appealing when the above KISS code should do the job. Actually they are all a source of potential bugs and will be harder to maintain, and harder for other team members to grok.

# A note on child bounds/clipping
I think this hit test outside of parent bounds fixes part of this issue. But what are we supposed to do about the size of our offset elements if they are bigger than tne parent? 

I always try to use `UnconstrainedBox` here, but it works in really bizarre ways. For example, if I use it here, I have to offset by 60px, instead of 40, to get the correct positioning (wat?)

**Comment by esDotDev** [comment]
Something like `IgnoreParentBounds` that respected parent positioning, but did not enforce any bounds on the size of the child, or the hit test bounds of the child would be perfect.

Then at any time, in any widget we can just make a Stack, wrap some thing in `IgnoreParentBounds`, and translate it anywhere we need.

**Comment by esDotDev** [comment]
Another example from a realworld use case we've had in the past.

A custom auto-complete in 63 lines, using the offset trick + inverting column order:
https://user-images.githubusercontent.com/736973/111886459-32c68180-8993-11eb-8149-b68cb9c10f5d.mp4
```
class SearchResultsSpike extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Column(
      // invert column direction so the SearchBox sits on top
      verticalDirection: VerticalDirection.up,
      children: [
        Flexible(child: Center(child: Text("PAGE CONTENT"))),
        ExpandingSearchBox(),
      ],
    );
  }
}

class ExpandingSearchBox extends StatelessWidget {
  final ValueNotifier<String> _textNotifier = ValueNotifier("");
  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _textNotifier,
      builder: (_, __) {
        String value = _textNotifier.value;
        return SizedBox(
          height: 60,
          child: Stack(
            children: [
              // Search Box
              TextFormField(onChanged: (value) => _textNotifier.value = value),
              // Search Suggestions Dropdown
              if (value != "")
                Transform.translate(
                  offset: Offset(0, 60),
                  child: OverflowBox(
                    alignment: Alignment.topLeft,
                    maxHeight: 400,
                    child: _SearchResults(query: value),
                  ),
                )
            ],
          ),
        );
      },
    );
  }
}

class _SearchResults extends StatelessWidget {
  final String query;

  const _SearchResults({Key? key, required this.query}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Container(
        width: double.infinity,
        color: Colors.grey.shade100,
        child: Column(
          children:
              List.generate(query.length, (i) => TextButton(child: Text("Some Suggestion... "), onPressed: () {})),
        ),
      ),
    );
  }
}
```

Note that only 6 lines here concern themselves with position of this dropdown in relation to the TextField, everything else is content. And we could drop that to 2 lines or so by making a little wrapper for Translate + OverflowBox:
```
verticalDirection: VerticalDirection.up,
... 
OverlayedWidget(offset: Offset(0, 60), maxHeight: 400, child: _SearchResults(query: value))
```

The simplicity is the win here.

**Comment by esDotDev** [comment]
Another really nice addition, which would remove the cludge around column direction, would be an OverlayedWidget:
```
return OverlayedWidget(child:  _SearchResults(query: value)); // gets sorted above all Widgets, but underneath Overlay
```
Allowing us to create overlay-like things, without all the complexity of Overlay. 

If things in the OverlayedWidget layer had unbounded hit tests, and sorted 1 layer above the normal widget layer, that could get us a lot of what we need. Though there are still use cases for thing sliding out from underneath other layers, so the general purpose `IgnoreParentPointerBounds` would still be needed.

**Comment by esDotDev** [comment]
Here is another example. A figma style selection box in 84 lines:

https://user-images.githubusercontent.com/736973/111888736-e84e0080-89a4-11eb-8178-37c2832994ae.mp4

This shows how we can have good encapsulation by not having the parent know or care about any controls in the negative margins of it's box. This box may have big fat controls for tablet, small controls for desktop, maybe it has panels that may or may not show up depending on content type, the parent does not, and should not, care about any of that, it just knows this box is 200w/200h and positioned at 200x/200y.

It also shows a technique, that I believe is impossible with Overlay, which is counter-scaling our overlayed content so it can maintain a consistent size. **Note in the video below that the border always appears to be 1px, and the handles are 10px, regardless of the scale of their parent.**

```
class _FigmaSelectionSpikeState extends State<FigmaSelectionSpike> {
  late final TransformationController _controller = TransformationController()..addListener(_handleTransform);

  @override
  Widget build(BuildContext context) {
    return InteractiveViewer(
        transformationController: _controller,
        boundaryMargin: EdgeInsets.all(double.infinity),
        child: Container(
          width: 1000,
          height: 1000,
          child: Artboard(parentScale: _controller.value.row0[0]),
        ));
  }

  void _handleTransform() => setState(() {});
}

class Artboard extends StatelessWidget {
  const Artboard({Key? key, required this.parentScale}) : super(key: key);
  final double parentScale;

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Container(color: Colors.grey.shade300),
        Transform.translate(
          // Position the Box at it's true position and size
          // We intentionally don't want to know about the "hanging" ui,
          // this is Box's problem.
          offset: Offset(200, 200),
          child: Container(
            width: 200,
            height: 200,
            child: ContentBox(isSelected: true, uiScale: parentScale),
          ),
        )
      ],
    );
  }
}

class ContentBox extends StatelessWidget {
  const ContentBox({Key? key, required this.isSelected, required this.uiScale}) : super(key: key);
  final bool isSelected;
  final double uiScale;

  @override
  Widget build(BuildContext context) {
    Widget scale(double value, Widget child) => Transform.scale(scale: value, child: child);
    double btnSize = 10;
    return Stack(
      clipBehavior: Clip.none,
      children: [
        // Background
        Container(color: Colors.white),

        // Box can overhang its UI, in any way it wants, no one else knows or cares
        if (isSelected) ...[
          Container(decoration: BoxDecoration(border: Border.all(width: 1 / uiScale, color: Colors.red))),
          Positioned(right: -btnSize / 2, top: -btnSize / 2, child: scale(1 / uiScale, _Handle(btnSize))),
          Positioned(left: -btnSize / 2, top: -btnSize / 2, child: scale(1 / uiScale, _Handle(btnSize))),
          Positioned(right: -btnSize / 2, bottom: -btnSize / 2, child: scale(1 / uiScale, _Handle(btnSize))),
          Positioned(left: -btnSize / 2, bottom: -btnSize / 2, child: scale(1 / uiScale, _Handle(btnSize))),
        ]
      ],
    );
  }
}

Widget _Handle(double size) => Container(
    width: size,
    height: size,
    decoration: BoxDecoration(color: Colors.white, border: Border.all(color: Colors.red)),
    child: TextButton(
      style: ButtonStyle(visualDensity: VisualDensity.compact, padding: MaterialStateProperty.all(EdgeInsets.zero)),
      onPressed: () {},
      child: Container(),
    ));

**Comment by rlch** [comment]
@esDotDev Thanks so much for taking the initiative here, I've been following this issue for a while now. I'm trying to implement a Figma-style layout in a production app, and can't perform hit-tests of widgets outside the bounds of my Stack within an InteractiveViewer. 

I've tried implementing overlays as a solution to no avail - fairly certain they can't be used to fit my use-case, and the one you've outlined above. 

Would love to see a solution to this implemented so that I don't have to dive into RenderObjects.

**Comment by modulovalue** [comment]
I have 2 points to add that speak against this feature:

1. Hit testing and painting is intuitive and easy to debug the way that it works right now. You paint in one direction, you hit test in the other. Overlays just lay on top of everything. I can't think of any exceptions to that. I believe this change would make that unnecessarily complicated to understand and debug. 
2. Having a stack with a bunch of widgets to implement complex UI elements is easy, but it is not efficient. It trades performance for ease of use.


In my opinion, the solution to this issue is already in this thread:
> Obviously that would in general be expensive [...] I think ~we would~ _you should_ still avoid doing this ever [...] since it's better to adjust the widget hierarchy to not have ancestors of the transformed widget be smaller than that widget [...]

That, and a lot more documentation, especially examples around RenderObjects and the Flow^ widget would solve this issue. 


--- 
### To those who want to implement a Figma like view
You will run into problems when your Stack contains widgets that are not visible. You need RenderObjects to conditionally create and destroy widgets. Or at least a Flow^ widget to conditionally paint a subset of them.

### InteractiveViewer is not made for the things that people use it for
This is a huge problem because it will disappoint those who try to use it for that.
Flutter supports tens of thousands of inflated widgets in one view, with hundreds of widgets in one viewport and can still reach 200FPS+ You will never be able to achieve this with a Stack (and an InteractiveViewer). This is not an issue with InteractiveViewer or with Widgets, it's just that all abstractions have a performance limit. People should be taught to not be afraid of RenderObjects.


> but the cost (which affects every app) might still be worth it?

No!

---

^ Not many people know of the Flow widget. It is a more powerful version of a Stack.  https://api.flutter.dev/flutter/widgets/Flow-class.html. It could solve some of the problems that people are facing here.

**Comment by esDotDev** [comment]
That Flow widget looks really cool, 18mths with Flutter and this is the first I've ever heard of it. 

1. From my experience what most devs find intuitive, is "if something is visible, and not covered by something else, I can click it", the idea of an invisible box that just cuts off your hit-testing is very non-intuitive. I mean, HTML, Unity, Flash they all act this way.

2. I don't see what that has to do with this issue. Most stacks we use render less than a dozen items, it wouldn't make sense to use a lower-level API, and we still want this flexibility in regards to hit testing.

3. I don't understand the argument against extra flexibility. You could simply not use the Widget if you choose.

>  it's just that all abstractions have a performance limit. 

While this is true, its also not meaningful. An abstraction layer can't simply throw up it's hands and say, not my fault. Not every abstraction layer would not have this massive delta in performance.

The idea that we can't simply position something, with an X/Y coords in a few lines, and expect decent performance, in 2021 does not ring true to me at all, knowing how powerful these devices all are.

**Comment by Hixie** [comment]
> 1. I mean, HTML, Unity, Flash they all act this way.

FWIW, HTML pays a huge perf cost for this, it's one of the reasons we avoided it in Flutter. (Dunno about the others. Given my experience in Tabletop Simulator, Unity hit testing is a huge mess, but I don't know if that's TTS or Unity!)

**Comment by modulovalue** [comment]
> In general, I don't understand the argument against extra flexibility. You could simply not use the Widget if you choose.

You might have accidentally skipped the third paragraph of Hixies issue description:

> [...] Having to check that flag every step of the walk would itself be not cheap, but the cost (which affects every app) might still be worth it?

Or in other words: Flutter would become slower for everybody if this feature would be implemented. That's why one could not just not use it.

---

@esDotDev you mention many things that are worth a comment. I can't put too much time into discussions like that right now, I'm sorry for that, I hope you don't mind that.

_If Flow isn't enough for you and you're not afraid to get dirty, Flutter has a hidden, powerful secret weapon stashed in its commit history.
Auto Layout! :)
https://github.com/flutter/flutter/pull/7350/commits/04c82a743d64ff94e171321e5b01d5f39e00791e_

**Comment by rlch** [comment]
For building a Figma-like view, if I opt for `Flow` is it okay to pair this with `InteractiveViewer`? Or should I suck it up and use `RenderObject`?

**Comment by modulovalue** [comment]
> For building a Figma-like view, if I opt for Flow is it okay to pair this with InteractiveViewer? Or should I suck it up and use RenderObject?

I'm sorry, but It's impossible to say from the outside, but maybe just go with the Flow and see if that would solve this issue for you first.

**Comment by Vardiak** [comment]
This issue really doesn't all come down to a Figma-style view. I've been developing complex UIs and this really comes up every time, even in simpler cases like a custom autocomplete widget (and no, Overlay is not always a viable solution). You often have to rethink your layout entirely for a little 5px margin that you would want to be hittable, this is really not practical. I don't have enough time to enumerate all the cases I've come across (thx @esDotDev) but really insist that this opt-in mechanism should be implemented.

**Comment by esDotDev** [comment]
> You might have accidentally skipped the third paragraph of Hixies issue description:

Ah, I did miss that point. Well we would need to quantify this before we can have any real discussion about it I think. Or hopefully come up with a solution that doesn't impose a tax on users who haven't opted in.

> This issue really doesn't all come down to a Figma-style view.

Exactly. It's a very handy tool for many tasks, it would be nice if we could use it. 

People proposing alternate solutions with Flow or Overlay, have to grapple with the fact that  my 3 examples above use <5 lines to do layout, and your proposals are 10-20x more code. This can't be hand-waved away and un-important, or devlopers being lazy, there are tons of tangible benefits to writing and maintaining less code.

**Comment by esDotDev** [comment]
> @esDotDev you mention many things that are worth a comment. I can't put too much time into discussions like that right now, I'm sorry for that, I hope you don't mind that.
> 
> _If Flow isn't enough for you and you're not afraid to get dirty, Flutter has a hidden, powerful secret weapon stashed in its commit history. Auto Layout! :) [04c82a7](https://github.com/flutter/flutter/commit/04c82a743d64ff94e171321e5b01d5f39e00791e)_

No worries at all, I have already learned loads from your 2 posts. Thanks!

**Comment by olof-dev** [comment]
I don't know that the opt-out mechanism described here is the way to go, but there was a request for use cases above so I'll add one: wanting a small tappable widget to 'slide out from behind' another widget (like a drawer that pulls out and doesn't take up 'physical space').

The natural way to implement this seemed to me to be to put the drawer behind the main widget in a Stack, and then have a SlideTransition animate it out when needed. This works just fine visually, but the sliding drawer doesn't detect hits because of the behaviour described here and elsewhere.

Using Overlay to get around this seems rather awkward to me, since one then needs to clip the drawer separately during the animation. Perhaps there are other nice ways of doing this, however, and I'd be happy to hear about them.

![screenshot](https://user-images.githubusercontent.com/66305960/113615318-ff9f1600-9653-11eb-8071-4e68ddf6131b.png)


```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with SingleTickerProviderStateMixin {
  late final AnimationController _controller = AnimationController(
    vsync: this,
    duration: const Duration(seconds: 1),
  )..repeat(reverse: true);
  late final Animation<Offset> _offset = Tween<Offset>(
    begin: Offset.zero,
    end: const Offset(0.0, 1.0),
  ).animate(CurvedAnimation(parent: _controller, curve: Curves.ease));

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Container(
        color: Colors.white,
        alignment: Alignment.center,
        child: Stack(
          alignment: Alignment.bottomCenter,
          children: [
            Positioned(
              bottom: 0.0,
              child: SlideTransition(
                position: _offset,
                child: GestureDetector(
                  onTap: () => print("Sliding drawer was tapped"),
                  child: Container(
                    color: Colors.blue[100],
                    child: Text("I slide out from behind", style: const TextStyle(fontSize: 16.0)),
                  ),
                ),
              ),
            ),
            Container(
              width: 300,
              height: 100,
              color: Colors.blue[300],
            ),
          ],
        ),
      ),
    );
  }
}
```

**Comment by olof-dev** [comment]
If we add a shadow to the main Container, which is fairly natural in the 'drawer' scenario, I don't quite see how to achieve this with Overlay (assuming we want to keep the main widget in the main tree). No doubt it wouldn't be a problem with a custom RenderObject, but that's a lot more code.

![screenshot_shadow](https://user-images.githubusercontent.com/66305960/113617853-290d7100-9657-11eb-8c0d-cc47f4de4bf5.png)

**Comment by Vardiak** [comment]
@olof-dev That's what I was talking about. This is the kind of example that is close to impossible to implement using an Overlay. This proposition could be accepted for this reason only.

**Comment by esDotDev** [comment]
Ya the shadow example in particular show-cases the flexibility of this technique. We will often get subtle designs like this from Design Team, and it is very important to them that a shadow sorts properly.

Essentially this technique gives developer more flexibility in sorting, whereas Overlay has only one sorting behavior and no ability to interleave the "follower" with the "target".

More flexible and less code!

**Comment by Vardiak** [comment]
@Hixie Can the name of this issue be changed? I feel like it is not gaining enough traction as other issues reporting a similar problem because it isn't obvious about what it means for beginner flutter developers.

**Comment by esDotDev** [comment]
Its also the 3rd or 4th post on this issue, the others had dozens of
upvotes that have basically been lost.

**Comment by Hixie** [comment]
The reality is this one might be closed as well. The suggestions in the original comment aren't really viable in practice, I just listed them to see if we could get some ideas for how to do it. But we haven't really gotten any.

**Comment by esDotDev** [comment]
So, is there any plan to address this limitation at all? We've listed 2 or 3 use cases, that while totally straightforward and expected from our designers POV, becomes literally impossible with Overlay. 

Specifically 
* sorting "pop-outs" with shadows as you would typically expect, 
* having a consistently sized overlay while the parent zooms, for example a 1px border around a box, as the box itself is scaled (seen in every design tool out there, Figma, XD, Sketch etc).
* having many overlays that should interleave the content (multiple boxes, with selection outlines, that are still sorted, again, like Figma, XD etc all do)

It's pretty frustrating, as flutter teams asks here" are then any use cases that would be simplified with this approach", many use cases are provided which clearly benefit from this approach, and then the issue is just closed anyways.

It seems to point to some flaw in how github issues are used to determine priorities. No matter how many issues are raised, or upvotes cast, it seems like this is just never going to change. I think at this point the community is just running out of energy pushing this, will give up, and Flutter just lives with this unfortunate limitation forever, which is extremely unfortunate imo.

**Comment by Hixie** [comment]
> * sorting "pop-outs" with shadows as you would typically expect,

This sounds like the same kind of thing as what we do with toolbars in text fields, which we do with Overlay. Unless you are referring to changing the order of the entries in the overlay? In which case I would probably recommend just making your own Overlay-like widget, or indeed just using a Stack (Overlay is basically just a Stack with an imperative API, after all), where you can control whatever you want.

I certainly would not recommend doing pop-outs using a hitTestSelf culling opt-out even if Flutter provided it. It would be really expensive to do it that way.


> * having a consistently sized overlay while the parent zooms, for example a 1px border around a box, as the box itself is scaled (seen in every design tool out there, Figma, XD, Sketch etc).

Not sure I understand what you mean here. Do you have a screenshot?


> * having many overlays that should interleave the content (multiple boxes, with selection outlines, that are still sorted, again, like Figma, XD etc all do)

Sorry I'm not familiar with Figma or XD. This sounds like something that I'd probably approach with custom RenderObjects though. I mean, any time you are really designing custom bespoke layouts, that's really what Flutter expects you to do.


> It's pretty frustrating, as flutter teams asks here "are there any use cases that would be simplified with this approach", many use cases are provided which clearly benefit from this approach, and then the issue is just closed anyways.

Oh I totally agree that there are use cases, that's why I filed this bug. There being use cases is a reason for us to look at a problem. There also have to be reasonable solutions if we're to implement them, though. This specific issue is narrowly focused on some potential solutions that could be *a* way to resolve a particular set of use cases, but there's lots of other ways unrelated to hit testing for solving those use cases.


> It seems to point to some flaw in how github issues are used to determine priorities. No matter how many issues are raised, or upvotes cast, it seems like this is just never going to change. I think at this point the community is just running out of energy pushing this, will give up, and Flutter just lives with this limitation forever, which is extremely unfortunate imo.

When we created Flutter, we made some design decisions that will bake in limitations. For example, we use Dart. We don't use TypeScript. So if the limitation is "Flutter requires Dart", then yes, we're unlikely to change that. Similarly we decided early on to never address vertical text requirements. This allowed us to make *massive* simplifications to our API design, but it means that certain Japanese use cases (for example) aren't something Flutter will ever be good at. Another example is that we decided to fix a huge performance problem with web browsers around hit testing, by doing hittest culling. I filed this bug because I thought it would be worth reexamining this if we could find efficient solutions that avoided this, but at the end of the day, we have not found more efficient ways to do hit testing that don't do culling.

Something being high priority doesn't mean it gets done, if there's no solution that addresses all the constraints.

Also, as far as I'm aware, all the use cases that have been brought up for removing hit test culling are quite possible if you design the application with hit test culling in mind. So it's not like this is preventing people from writing apps, it's just requiring that people design their apps a little differently than they might do when designing them for other systems. (Notably the web; most other systems actually do do hit test culling as well. It may be less obvious in some cases because in most systems the tree is extremely shallow.)


> Worth mentioning that this "expensive" hit test, to date, has never been quantified in any way. Which seems a rather important first step to evaluating anything...

An easy way to see how expensive hit testing is is to profile Flutter desktop applications when you casually move the mouse over the window. It's almost always the top of the profile. We're spending way more time hit testing today than I'm comfortable with, frankly. (I wish we could just remove hover support, but sadly that's a non-starter on desktop.) Making hit testing more expensive (by moving it from O(N) in depth of tree to O(M) in nodes in the tree, which can be multiple orders of magnitude bigger) is just not practical.

You could easily test this yourself. Remove the culling logic from the framework, and compare how long hit-testing takes before and after.

If you have specific use cases that you can't solve, or for which you think the framework should provide a solution (as opposed to a package providing the solution), then you should feel free to to file bugs on those use cases. I would recommend not trying to describe the preferred solution as part of describing the problem, though, as you will get more traction if you explain that you can't figure out how to do something than if you say "this particular API is preventing me from doing something". (Because it's easy to say "no it isn't" to the second, but hard to deny the first without providing an example.)

**Comment by Vardiak** [comment]
> An easy way to see how expensive hit testing is is to profile Flutter desktop applications when you casually move the mouse over the window. It's almost always the top of the profile. We're spending way more time hit testing today than I'm comfortable with, frankly. (I wish we could just remove hover support, but sadly that's a non-starter on desktop.) Making hit testing more expensive (by moving it from O(N) in depth of tree to O(M) in nodes in the tree, which can be multiple orders of magnitude bigger) is just not practical.

I'm having trouble understanding this. We are asking for a way to make a specific Widget hittable globally, without needing it to be in its parents bounds. Wouldn't this require only to loop on those few specific Widgets to check if it gets hit? I don't understand how this could be O(M) with M = number of nodes on the tree.

**Comment by esDotDev** [comment]
"So it's not like this is preventing people from writing apps" - In a world without timelines, and requirements that are fully known in advance, this might be true, but this world doesn't exist.

In reality, when things are slow cludgy to implement, or have to be "planned for" months in advance to not be a total mess, the feature just gets cut, or the design gets downgraded. 

I understand that not everything is possible, but this is a basic ask from an end user perspective, it's not analogous to switching to dart, or asking for vertical text. We haven't even done the basic step of testing these hit tests to see the magnitude of what we're talking about, so I don't see how any of the proposed solutions can really be dismissed. What exactly is "expensive"?

**Comment by Vardiak** [comment]
> In reality, when things are slow cludgy to implement, or have to be "planned for" months in advance to not be a total mess, the feature just gets cut, or the design gets downgraded. So this very much does stop features and design from being built, if not entire apps.

This. I cannot count the number of time I had to ask the designer to rethink the UI because it was just too hard to do it in Flutter whereas the same exact feature is implemented on the web very easily. Flutter plays nicely with Material, but I work with complex UIs and this specific issue is the only thing that makes integration hard in my opinion.

> Not really sure where to start here, there's 10's of thousands of lines of code in the FlutterSDK. Happy to play with it if you can point me in the right direction.

I would also be happy to help, a detailed explanation of how this could be done would be much appreciated! I have multiple ideas on how to do it efficiently in my head but I don't know Flutter architecture enough to do it on my own.

**Comment by esDotDev** [comment]
This issue still seems to be open, and accurately describes the problem, so I don't think there's need for a new issue. 

While tweaks to Overlay would be nice, to support some edge cases, what we really are missing is a simple way to expand the hit area x pixels: https://github.com/flutter/flutter/issues/31728

The bummer is that this issue should have more like 150+ likes, if you combine all the issues that are complaining about the same thing.

fwiw, there is a package now that seems to do the job, referenced in that post:
https://pub.dev/packages/expand_tap_area

@Hixie Maybe this could inspire some ideas within the team on a way to tack this? Seems to use a custom `GestureArenaMember` with some negative offsets.

Notably, within the source code, we find this comment:
 _// If there is someone who knows how to extend the hitarea above the parent rect, so it is not
// Clipped, please add that here. **This would be really helpful.**_
https://github.com/malikwang/expand_tap_area/blob/master/lib/src/expand_tap_area.dart

**Comment by Hixie** [comment]
@Vardiak 
> > An easy way to see how expensive hit testing is is to profile Flutter desktop applications when you casually move the mouse over the window. It's almost always the top of the profile. We're spending way more time hit testing today than I'm comfortable with, frankly. (I wish we could just remove hover support, but sadly that's a non-starter on desktop.) Making hit testing more expensive (by moving it from O(N) in depth of tree to O(M) in nodes in the tree, which can be multiple orders of magnitude bigger) is just not practical.
> 
> I'm having trouble understanding this. We are asking for a way to make a specific Widget hittable globally, without needing it to be in its parents bounds. Wouldn't this require only to loop on those few specific Widgets to check if it gets hit?

The difficulty is that without walking the tree, you don't know which widget is on top when two of your specific widgets overlap.

If you don't care (e.g. because you know your widgets never overlap) you can do this today by hooking into the pointer router at the top level and, checking the pointer against the widget's bounds. No need for anything new in the framework to do that.


> This. I cannot count the number of time I had to ask the designer to rethink the UI because it was just too hard to do it in Flutter whereas the same exact feature is implemented on the web very easily. Flutter plays nicely with Material, but I work with complex UIs and this specific issue is the only thing that makes integration hard in my opinion.

If there are specific UIs that are too hard to do in Flutter, please file a bug describing that UI. (Don't file a bug asking for a specific solution, like "expand hit-test areas", file a bug for the specific UI you're trying to build.)


> > Not really sure where to start here, there's 10's of thousands of lines of code in the FlutterSDK. Happy to play with it if you can point me in the right direction.
> 
> I would also be happy to help, a detailed explanation of how this could be done would be much appreciated! I have multiple ideas on how to do it efficiently in my head but I don't know Flutter architecture enough to do it on my own.

The hit testing system is documented pretty heavily, see e.g. RenderBox in the API docs.

The code that does the filtering I'm talking about is typically in `hitTest`, e.g. `RenderBox.hitTest` (that one provides the default implementation for most RenderBox subclasses). If you look at that function you'll see it only calls `hitTestChildren` if the `position` is inside `size`. Remove that check, and it'll walk the entire tree and all children will check their bounds without paying attention to the parent.

To benchmark it, do that in some complicated application, and then time how long it takes to call `binding.hitTest` thousands of times with random offsets, and compare to the performance with the check present.

**Comment by marcglasberg** [comment]
I am late to this discussion, and this proposal may make no sense at all, but how about we have these 2 widgets:

1) A new `HitTestAll` widget, where all widgets below it would call their children hit test. So the developer could choose the specific subtree they want to pay the performance hit.

2) The GestureDetector widget could have a `hitMargin` parameter to expand its hit test area.

So for example, if I want some text in a row to have a 20 pixel larger click-area I could do:

```
HitTestAll(
  child: Row(
      children: [
          Text('abc'),
          Text('123'),
          GestureDetector(
               onTap: onTap,
               hitMargin: EdgeInsets.all(20),
               child: Text('Click me'))
```

**Comment by Hixie** [comment]
> 1. A new `HitTestAll` widget, where all widgets below it would call their children hit test without further culling. So the developer could choose the specific subtree they want to pay the performance hit.

The problem is that it is not possible (as far as I can tell) to create such a widget without taking a performance hit even if it's not used, because every widget would need to check whether it was inside such a subtree all the time (even when not in such a subtree).


> 2. The GestureDetector widget could have a `hitMargin` parameter to expand its hit test area.

Similarly here, it doesn't seem to me to be possible to implement this without it affecting every widget (technically every render object) because it requires all the hit test logic for nodes above the GestureDetector in the tree to cooperate. Currently, the gesture detector's hit test area is literally the biggest it can be according to its parent. So without cooperation from the parent, it can't be made bigger.

That said, I could be wrong. If there's a way to implement these widgets, I would encourage you to create a package that contains them. It would be a popular package if this issue is anything to go by. :-)

**Comment by marcglasberg** [comment]
@Hixie @esDotDev 

> because every widget would need to check whether it was inside such a subtree all the time (even when not in such a subtree).

Can't we send this information (if it's inside such a subtree or not) down, through the `hitTest` method? 

This is the current `hitTest` signature:

```
bool hitTest(BoxHitTestResult result, { required Offset position });
```

It could have an `expandArea` boolean:

```
bool hitTest(BoxHitTestResult result, { required Offset position, required bool expandArea});
    if (expandArea || _size.contains(position)) {
      if (hitTestChildren(result, position: position, expandArea: expandArea) || hitTestSelf(position)) {
        result.add(BoxHitTestEntry(this, position));
        return true;
      }}}
```

The expandArea parameter would be `false` until it encountered the `HitTestAll` widget, which would simply change it to `expandArea: true`: 

```
bool hitTest(BoxHitTestResult result, { required Offset position, required bool expandArea}) {
    if (expandArea || _size.contains(position)) {
      if (hitTestChildren(result, position: position, expandArea: true) || hitTestSelf(position)) {
        result.add(BoxHitTestEntry(this, position));
        return true;
      }}}
```

One can't really create a package with these features because it means changing the `hitTest` method.

**Comment by Hixie** [comment]
What you describe is exactly what I am referring to when I say "every widget would need to check whether it was inside such a subtree all the time".

**Comment by marcglasberg** [comment]
@Hixie Ah, ok. I was thinking you meant that each widget would have to go up the tree to search for an ancestor, somehow. That could potentially be slow.

What I've described means having to pass an extra parameter to the method that already exists and checking a boolean for each widget. This:

```
if (expandArea || _size.contains(position)) 
```

Instead of this:
```
if (_size.contains(position)) 
```

Even in a very large tree, if we have, say 10,000 widgets to check, I'd say that's extremely fast. I can create some code to actually measure it, if you want. 

Also, I'd say that even if Flutter doesn't want to provide a widget like the `HitTestAll` widget I've described above, changing the `hitTest` method would allow package authors to create widgets that opt out of the hit test culling, just as you have suggested.

**Comment by Hixie** [comment]
Sure, that one boolean check wouldn't make a noticeable difference. However, the general policy of whether or not we add such features applies to many more places than just this check. If we have the general policy of adding features like this, then there would be many, many more places where we would add such checks. Together, they _would_ be noticeable.

So we have three choices: add this feature because we have a policy of adding such features, add this feature in violation of our policy of not adding such features because it's an exception that deserves special consideration, and not add this feature because we have a policy of not adding such features.

It's only by being really rigorous in applying such policies that we can maintain performance across the board., so our bar for violating the policy is a very high one. Considering whether this meets that bar is why I filed this bug in the first place, but so far I don't think this is an unusually compelling case given that there are workarounds available (albeit unpleasant ones).

**Comment by esDotDev** [comment]
Everyone will have their own list of what deserves special consideration, but I don't think you get much more integral than hit testing. It's the foundation of virtually all interaction with the devices. It is certainly right at the very top of my list for "Frustrating pain points in flutter". The community seems to have put in a lot of effort to create code examples to demonstrate the issue, and poured in hundreds of votes over the years.

I think this raises an interesting question of how we get past impasses like this. Where the community clearly wants/needs something, but the team seems reluctant to move on it, and we just stall out. I wonder how it can really be decided what "deserves special consideration"?

**Comment by marcglasberg** [comment]
@Hixie Well, I don't know the policy you are talking about, so I can't comment on how this issue fix would affect it. 

It's just that I keep seeing people complaining about this specific issue. For example: https://github.com/flutter/flutter/issues/19445 and also this https://github.com/marcglasberg/align_positioned/issues/7

And also these workarounds to expand the tap area make my code ugly. I create some really nice clean code and then I have to make it ugly to expand the click-area. It just feels unnatural.

And there is at least one use case I can't fix with current workarounds. If we have some text and in the middle of it there's a link. I'd like to expand the link area, but I don't know where it's going to be printed because it's a very complex calculation I can't perform myself.

Please note, there is already a package that would start working properly if the `hitTest` method were changed:  https://pub.dev/packages/expand_tap_area

**Comment by Hixie** [comment]
I don't think there's any disagreement that this is something we'd like to make better. I mean, I filed this bug!

We're just trying to find a way to fix it that doesn't involve a performance hit.

> I think this raises an interesting question of how we get past impasses like this. Where the community clearly wants/needs something, but the team seems reluctant to move on it, and we just stall out. I wonder how it can really be decided what "deserves special consideration"?

The community is not a monolithic entity (nor is the team). It's not really useful to frame this as an "us vs them" situation. People care very much about performance too. People care very much about how Flutter's widgets are self-contained and how the protocol between widgets and render objects is simple. We're all in this together, trying to find a design that can address people's needs without compromising on these other priorities.

**Comment by marcglasberg** [comment]
Well, I certainly agree those things you mentioned are important. In any case, if the Flutter team decides that the proposed change in the `hitTest` method would be enough to address this issue, let me know if you need measurements on how it would affect performance. Thanks.

**Comment by csells** [comment]
@marcglasberg if you find yourself motivated, it would be interested to see those perf numbers if for no other reason than to convince ourselves that we're right not to add this feature.

**Comment by marcglasberg** [comment]
@csells Well, I would fell motivated only if the proposed change in the hitTest method was considered the solution to the issue, pending only the numbers (the performance hit). But, if as you say the team has already decided not to add the feature for other reasons, it's ok for me.

**Comment by esDotDev** [comment]
Its really quite hard to implement these sorts of deep changes to the framework and then accurately measure the cost. It's a tough task to start modifying such a complex system you are not intimately familiar with.

Also, once implemented, measuring the performance of such a tiny change also seems like it would be extremely challenging if you're not deeply knowledgeable on benchmarking in Flutter. Even in simple loops, with 10m iterations, I struggle to measure any consistent cost to a simple bool check. Either the compiler is optimizing things out, or the dart is stupid fast when it comes to comparing bools. 

The team is really in a much better position to run some tests, and come back with a number for the cost here, and then we would have real numbers to discuss, vs some undefined idea of "cost". 

For example, if it turns out that 10,000 bool checks, are <.01ms on a mid-tier ARM processor, would this discussion be moot? It seems like the order of magnitude should matter here, and be the first data point we need before any potential solution can be considered.

**Comment by marcglasberg** [comment]
@esDotDev I don't think it's that hard. If you copy the Stack and RenderStack code and modify the hitTest method, then you can create an arbitrarily large wide/deep random tree of those stacks inside of stacks. Then you hit test random locations a million times comparing the time it takes for each different `hitTest` method.

I'd compare those:

The current one:
```
bool hitTest(BoxHitTestResult result, { required Offset position});
    if (_size.contains(position) {
      if (hitTestChildren(result, position: position) || hitTestSelf(position)) {
        result.add(BoxHitTestEntry(this, position));
        return true;
      }}}
```

No culling at all:
```
bool hitTest(BoxHitTestResult result, { required Offset position});
      if (hitTestChildren(result, position: position) || hitTestSelf(position)) {
        result.add(BoxHitTestEntry(this, position));
        return true;
      }}
```

Selective culling (the proposed solution)
```
bool hitTest(BoxHitTestResult result, { required Offset position, required bool expandArea});
    if (expandArea || _size.contains(position)) {
      if (hitTestChildren(result, position: position, expandArea: expandArea) || hitTestSelf(position)) {
        result.add(BoxHitTestEntry(this, position));
        return true;
      }}}
```

**Comment by shrouxm** [comment]
hi all, going to go ahead and poke this discussion again. i've been looking into this a bit because i'm trying to create some humble dropdowns for a desktop app (see some of my misadventures above). i potentially have a solution that will work for my use case, but i wanted to see what some more knowledgeable flutter folks think and whether there's a way to expand it to other use cases. i'm unclear on both the performance implications and soundness of the approach because i'm still pretty hazy on how the RenderObject life cycle and paint phase work.

the solution is sort of inspired by the CompositedTransform{Target,Follower} widgets and along the lines of what @Hixie mentioned in the original comment of this thread:
> Another option might be to have a way for a widget to hook into the hit testing of a far ancestor (e.g. via an inherited widget).

basically there's an ancestor RenderProxyBox which gets a handle to some deep children RenderProxyBoxes, and the ancestor does painting and/or hitTesting instead of the children.

<details>
  <summary>anyway, here's some code:</summary>

```dart
class DeferredPaintLink extends ChangeNotifier {
  final List<RenderBox> _renderBoxes = [];
  List<RenderBox> get renderBoxes {
    return UnmodifiableListView(_renderBoxes);
  }

  void add(RenderBox deferredPainter) {
    if (!_renderBoxes.contains(deferredPainter)) {
      _renderBoxes.add(deferredPainter);
      notifyListeners();
    }
  }

  void remove(RenderBox deferredPainter) {
    if (_renderBoxes.contains(deferredPainter)) {
      _renderBoxes.remove(deferredPainter);
      notifyListeners();
    }
  }

  DeferredPaintLink();
}

class DeferredPainter extends SingleChildRenderObjectWidget {
  DeferredPainter({required this.link, Widget? child, Key? key}) : super(child: child, key: key);

  final DeferredPaintLink link;

  @override
  RenderObject createRenderObject(BuildContext context) {
    return RenderDeferredPainter(link);
  }

  @override
  void updateRenderObject(BuildContext context, RenderDeferredPainter object) {
    object.link = link;
  }
}

class RenderDeferredPainter extends RenderProxyBox {
  RenderDeferredPainter(DeferredPaintLink link, {RenderBox? child}) : super(child) {
    this.link = link;
  }

  bool _linked = false;

  DeferredPaintLink? _link;
  DeferredPaintLink get link => _link!;
  set link(DeferredPaintLink link) {
    if (_linked) {
      link.remove(this.child!);
      _linked = false;
    }
    _link = link;
    if (_link != null && this.child != null) {
      link.add(this.child!);
      _linked = true;
    }
  }

  @override
  set child(RenderBox? child) {
    if (_linked) {
      link.remove(this.child!);
      _linked = false;
    }
    super.child = child;
    if (_link != null && this.child != null) {
      link.add(this.child!);
      _linked = true;
    }
  }

  void attach(covariant PipelineOwner owner) {
    super.attach(owner);
    if (!_linked && _link != null && this.child != null) {
      link.add(this.child!);
    }
  }

  void detach() {
    if (_linked) {
      link.remove(this.child!);
    }
    super.detach();
  }

  @override
  bool hitTest(BoxHitTestResult result, {required Offset position}) {
    return false;
  }

  @override
  void paint(PaintingContext context, Offset offset) {}
}

class DeferredPaintTarget extends SingleChildRenderObjectWidget {
  DeferredPaintTarget({required this.link, Widget? child, Key? key})
      : super(child: child, key: key);

  final DeferredPaintLink link;

  @override
  RenderObject createRenderObject(BuildContext context) {
    return RenderDeferredPaintTarget(link);
  }

  @override
  void updateRenderObject(BuildContext context, RenderDeferredPaintTarget object) {
    object.link = link;
  }
}

class RenderDeferredPaintTarget extends RenderProxyBox {
  RenderDeferredPaintTarget(DeferredPaintLink link, [RenderBox? child]) : super(child) {
    this.link = link;
  }

  DeferredPaintLink? _link;
  DeferredPaintLink get link => _link!;
  set link(DeferredPaintLink link) {
    if (_link != null) {
      _link!.removeListener(this.markNeedsPaint);
    }
    _link = link;
    this.link.addListener(this.markNeedsPaint);
    markNeedsPaint();
  }

  @override
  bool hitTest(BoxHitTestResult result, {required Offset position}) {
    for (final renderBox in link.renderBoxes.reversed) {
      final hit = result.addWithPaintTransform(
        transform: renderBox.getTransformTo(this),
        position: position,
        hitTest: (BoxHitTestResult result, Offset? position) {
          return renderBox.hitTest(result, position: position!);
        },
      );
      if (hit) {
        return true;
      }
    }
    return child?.hitTest(result, position: position) ?? false;
  }

  @override
  void paint(PaintingContext context, Offset offset) {
    super.paint(context, offset);
    for (final renderBox in link.renderBoxes) {
      context.paintChild(renderBox, renderBox.localToGlobal(Offset.zero, ancestor: this) + offset);
    }
  }
}
```
</details>

i have it set up so the ancestor paints the children last and hit tests first because i'm trying to make dropdowns, but you could for example have the children paint themselves, and then have the ancestor hit test them last if they're meant to be hiding behind the rest of the UI. the DeferredPaintLink object could of course be packaged with some widget inheritance to make it an implementation detail, i just wanted to keep things barebones for discussion's sake.

<details>
  <summary>and here's a sample usage:</summary>

```dart
void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => MyAppState();
}

class MyAppState extends State<MyApp> {
  final link = DeferredPaintLink();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: DeferredPaintTarget(
          link: link,
          child: Column(
            children: [
              Container(
                height: 200,
                child: OverflowBox(
                  alignment: Alignment.topCenter,
                  maxHeight: 400,
                  child: Column(
                    children: [
                      Container(
                        height: 200,
                        child: HitTestWidget(off: Colors.deepOrange, on: Colors.orange),
                      ),
                      DeferredPainter(
                        link: link,
                        child: Container(
                          height: 100,
                          child: HitTestWidget(off: Colors.green, on: Colors.lightGreen),
                        ),
                      ),
                      DeferredPainter(
                        link: link,
                        child: Container(
                          height: 100,
                          child: HitTestWidget(off: Colors.blue, on: Colors.lightBlue),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
              Flexible(child: HitTestWidget(off: Colors.amber, on: Colors.yellow)),
            ],
          ),
        ),
      ),
    );
  }
}

class HitTestWidget extends StatefulWidget {
  final Color off;
  final Color on;

  const HitTestWidget({Key? key, required this.off, required this.on}) : super(key: key);

  @override
  State<StatefulWidget> createState() => HitTestWidgetState();
}

class HitTestWidgetState extends State<HitTestWidget> {
  bool flipped = false;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () => setState(() => flipped = !flipped),
      child: Container(
        color: flipped ? widget.on : widget.off,
      ),
    );
  }
}
```
</details>

you can remove the DeferredPainters from the sample usage to see how they affect the behaviour. you could also customize whether the ancestor or children do either of the hit testing or painting and in what order to achieve various useful and/or nonsensical effects.

would something like this be useful for anyone else running into this issue? does this have subtle issues which make it a nonviable approach?

thanks for your time, anyone who read all this!

**Comment by esDotDev** [comment]
@shrouxm this looks really neat. So basically somehow you have some object at the top of the tree that is able to perform hit-tests for any  renderObject further down the tree, if given the `link`. 

Basically the renderObjects get painted and hit tested on a seperate layer, which sortof acts like a more flexible easier to use Overlay?

Would this still work if I wanted to do something basic, like offset an icon btn slightly outside my bounds?
```
Column(children: [
  DeferredPainter(child: Transform.offset(child: IconBtn(), offset: Offset(-5, 0))),
])
```

**Comment by esDotDev** [comment]
The ability to not render on top seems really interesting too. Sometimes you have a panel, with a shadow, and you want your dropdown to slide out underneath the shadow. This would support that too I guess, as I could render where I am, but have hit testing performed at the top of the tree?
![image](https://user-images.githubusercontent.com/736973/141230771-92a506bd-8219-49aa-8e5e-4aa1fb507b4d.png)

**Comment by esDotDev** [comment]
@shrouxm So far this looks really promising, any idea how I could tweak it to render in place so it can stay under the shadow?

For example, this works nicely, but I need it to sit under the shadow of the sibling container:
```
Stack(
  clipBehavior: Clip.none,
  children: [
    Positioned(
      bottom: -30,
      child: DeferredPainter(
        link: link,
        child: _SomeBtn(),
      ),
    ),
    Container(
        width: 100,
        height: 100,
        decoration: BoxDecoration(color: Colors.green, boxShadow: [
          BoxShadow(color: Colors.black.withOpacity(1), blurRadius: 4, spreadRadius: 4),
        ])),
  ],
),
```

https://user-images.githubusercontent.com/736973/141722886-2fc5b719-7631-415e-8910-0b565d86d128.mp4


@Hixie any thoughts on this approach in general? It seems usable, and only incurs performance hits on widgets that ask for it. Basically we have an ancestor widget, that can perform hitTesting and (optionally) rendering for us, allowing us to use some parent as our hit test bounds, but also keep our widget in the display tree if desired, so it maintains existing display order (which serves use cases that overlay does not support)

**Comment by esDotDev** [comment]
Also seems like `DeferredPaintTarget` could just have it's own internal link, and `DeferredPainter` just looks upwards? Link could become optional?  One single overlay-like surface at the root of the tree is the common case?

**Comment by esDotDev** [comment]
Nevermind @shrouxm I figured it out, here's an alternate version that adds a 'drawOnTop` value to the painters:
https://gist.github.com/esDotDev/0f783f1f4ddc318c967328b48bfd3199

With that we get this!
https://user-images.githubusercontent.com/736973/141724789-23731dd2-6350-4b7b-8b8e-31433bc626b0.mp4

**Comment by esDotDev** [comment]
@shrouxm I hope you don't mind but I made some small changes and packaged this up: 
https://pub.dev/packages/defer_pointer

As far as I can tell it solves all the use cases I can think of:

Offset buttons outside of a stack:

https://user-images.githubusercontent.com/736973/141850225-234f2ac7-afd6-4781-b58c-f78b39f445b2.mp4

A classic desktop style dropdown menu:

https://user-images.githubusercontent.com/736973/141850218-edb47e3a-1422-4471-902c-b23b3e181dfc.mp4

A `Flow` based widget (which for some reason, needs to use 2 `DeferHit` widgets, not sure why but it works!)

https://user-images.githubusercontent.com/736973/141850216-aae2434a-f679-446b-9e5c-ddcc96399afa.mp4

A auto-complete example:

https://user-images.githubusercontent.com/736973/141850211-b392f5e4-68d4-4ac2-b103-dc60133e377b.mp4

**Comment by Hixie** [comment]
In general it's plausible but note that linking between nodes in these trees is very bug-prone, you have to be really really careful about managing lifetimes and so on.

**Comment by shrouxm** [comment]
@esDotDev i don't mind that you packaged it, although it would be nice if you left a note crediting where the idea came from as the core logic is fairly identical to what i posted. i was thinking of different ways to package it and didn't really settle on a convenient/flexible API. you could for example do full blown z-indexes by making the link hold a SplayTreeMap of lists of render objects indexed over some comparable type. but i don't really think z-indexes are a good idea. 

i've only found one bug while using this code since i posted it, the `markNeedsPaint()` method of the child render object should also call `markNeedsPaint()` on the parent (if the parent is painting the child and not just doing hit testing,). otherwise a self-contained update of the child render object won't actually repaint on the screen.

**Comment by esDotDev** [comment]
@shrouxm attribution added, thanks for the idea! lmk if you want me to change the text, or happy to take any PR's you have

**Comment by esDotDev** [comment]
> In general it's plausible but note that linking between nodes in these trees is very bug-prone, you have to be really really careful about managing lifetimes and so on.

Thanks for the headsup, can you think of any problematic edge cases I should test? Or just expand on the potential issues? I'm not very familiar with `SingleChildRenderObjectWidget` lifecycle. Currently descendants just use attach/detach to add/remove themselves from the link.
https://github.com/gskinnerTeam/flutter-defer-pointer/blob/89fd01b9c1cb97b33c1897f7b3d1f70e20688a00/lib/defer_pointer.dart#L80

**Comment by Vardiak** [comment]
@shrouxm @esDotDev This is amazing! Exactly what I was looking for. I think this should be added to the framework, and maintained as a non-hacky way to do hit testing.

**Comment by esDotDev** [comment]
I agree with it having a nice place in the framework... `IgnorePointer`, `AbsorbPointer`, `DeferPointer` LGTM!

**Comment by Vardiak** [comment]
> I agree with it having a nice place in the framework... `IgnorePointer`, `AbsorbPointer`, `DeferPointer` LGTM!

About the name: I feel like it wouldn't be clear for and end user what it does, even though it is a good technical term. This is basically global hit testing, and the name should suggest that idea. This would be even simpler if the `DeferredPointerHandler` is moved inside the `WidgetsApp` widget (as I understand, there is no performance impact of placing the `DeferredPointerHandler` higher up the tree).

**Comment by shrouxm** [comment]
the performance implication of putting it further up would be unnecessary repaints affecting a larger portion of the app. as far as i understand, inserting a repaint boundary between parent/handler and its primary child should solve this but i haven't tested.

maybe another way to think about it could be via layers? since the repaint boundary would just be creating another layer. i'm envisioning the parent widget accepting a list of keys which will become labels under the scope of that widget for different semantic layers (e.g. maybe a layer for dropdowns, maybe a layer for widgets which only subtly extend beyond their bounds but should lie underneath the dropdowns, maybe a layer below the primary one? not sure what a use case for that one would be 🤔). then child widgets can choose which semantic layer to live & separately hit test on.

probably overwrought though, and as compared to the current `Overlay`'s way of doing things, it would make it more difficult to create re-usable dropdowns etc

anyway the behaviour of this widget is much closer to how i expected overlays to work than the current `Overlay` widget, and my impression is that people generally find `Overlay` fairly unintuitive to use (although i think the underlying mechanism of it is probably easier to wrap your head around than this thingy). so it would be pretty cool to upstream something like this; i'd be happy to workshop an API and hack on a PR if there's interest from any maintainers!

**Comment by esDotDev** [comment]
Well that is the nice thing actually, it's not global, you can stick `DeferredPointerHandler` at any level you need that works for you. So far I'm finding that is often just 1 or 2 levels up the stack. And you can have many of them, as they are really just interested in their immediate parent.

fwiw, I had it named `DeferHit` for most of  development, but just based the name to more closely match the existing `____Pointer` widgets.

**Comment by esDotDev** [comment]
In terms of sorting, I think we could get really far with a simple `DeferHit.paintOrder` property or something, now you can defer painting to some ancestor, and also manually control sorting within that layer if needed. That should serve most use cases I would think.

**Comment by fzyzcjy** [comment]
Hi, is there any updates?

**Comment by ishiWaldemarSV** [comment]
Hi there, 

**Are there any updates on adding a "hitSlop" (RN language) to Flutter?**
Coming from React Native, this limitation seems surprising. 

Yes, I created some mess with hitSlope in the past. There is a learning curve to that. 

On the other hand it allows to achieve vetera elegant UI, where the border doesn't have to be huge, the icon small and the touchable area easy to click. 


> "So it's not like this is preventing people from writing apps" - In a world without timelines, and requirements that are fully known in advance, and every team member is a 3 year flutter veteran, this might be true, but this world doesn't exist.

Bravo for the reply.

**Comment by Shahmirzali-Huseynov** [comment]
Hi, are there any updates?

**Comment by Vardiak** [comment]
> Hi, are there any updates?

You can use this library which does everything you want https://pub.dev/packages/defer_pointer

**Comment by Shahmirzali-Huseynov** [comment]
> > Hi, are there any updates?
> 
> You can use this library which does everything you want https://pub.dev/packages/defer_pointer

I tried this package, but Unfortunately,  it's not worked in Gridview

**Comment by gnprice** [comment]
The reference to React Native's `hitSlop` (https://github.com/flutter/flutter/issues/75747#issuecomment-1253366545) makes for an interesting point of comparison, because RN `hitSlop` carries a crucial limitation:

By setting `hitSlop`, you can let an RN component receive touches that happen outside its own bounds. But no matter how big you set `hitSlop`, it will never receive touches that happen outside its *parent's* bounds. In [the RN docs'](https://reactnative.dev/docs/pressable) terms:
> The touch area never extends past the parent view bounds and the Z-index of sibling views always takes precedence if a touch hits two overlapping views.

I imagine the motivation for that limitation is exactly to mitigate the type of performance concern that @Hixie has explained in this thread: it means that when a parent component gets a potential hit, if it's outside its own bounds then it never has to pass it on to its children, regardless of what those children are trying to do with `hitSlop`.

In my own experience with React Native, the restriction to the parent's bounds feels kind of annoyingly arbitrary, but it doesn't pose much of an obstacle in practice. So I think `hitSlop`, even with that limitation, solves a lot of the use cases this issue is about.

In Flutter terms, I think the analogous design would look something like:
* A `HitSlop` widget configures a `HitSlopRenderObject`.
* For layout and paint, a `HitSlopRenderObject` defers to its child.
* For hit-testing, a `HitSlopRenderObject` checks whether the position is in, not its actual size, but an expanded rectangle around its size, expanded by the widget's hit slop value. (Which would perhaps be an `EdgeInsetsGeometry`, just like `Padding.padding`.)

I believe that would function a lot like RN's `hitSlop`. It'd have the same arbitrary-feeling limitation, but would solve much the same set of use cases. And, crucially, it would cost nothing when you aren't using it, because no other widgets would have to change or gain an extra case to have to check.

**Comment by spkersten** [comment]
@gnprice in my project we have a widget like that. It works (in some cases) but it is also error prone: Very easy for someone to wrap the widget in a Row that hugs in vertically (or Opacity, ...) and you've reduced the hit test area.

**Comment by andersonmendesdev** [comment]
> > > Hi, are there any updates?
> > 
> > 
> > You can use this library which does everything you want https://pub.dev/packages/defer_pointer
> 
> I tried this package, but Unfortunately, it's not worked in Gridview

It solved it for me, I put the DeferredPointerHandler at the highest point of the tree on the screen in question and where it needed to be touched I added the DeferPointer and the capture of the touch outside the stack went well.

**Comment by Shahmirzali-Huseynov** [comment]
@andersonmendesdev can you share some code?

**Comment by Wdestroier** [comment]
> > Hi, are there any updates?
>
> You can use this library which does everything you want https://pub.dev/packages/defer_pointer

@Vardiak Thank you so much! :heart:

@shahmirzali49 It worked for me on second try. You may need a Builder between DeferredPointerHandler and DeferPointer, read the [documentation](https://api.flutter.dev/flutter/widgets/Builder-class.html) to understand.  That's how InheritedWidgets, such as Provider, work afaik.

The following works:
MaterialApp -> Scaffold -> DeferredPointerHandler -> LayoutBuilder -> Stack -> Transform -> DeferPointer -> Container
The following doesn't work:
MaterialApp -> Scaffold -> LayoutBuilder -> DeferredPointerHandler -> Stack -> Transform -> DeferPointer -> Container

**Comment by gikwegbu** [comment]
Hi Everyone,
Incase we are still facing this issue, I had a workaround for it.

1. Since the Clip.none, literally stops you from clicking any area outside the Stack, I added a SizedBox as the parent of the Stack, then made the height to house all my stacked designs.

2. For simplicity sake, I made a modular method for each stacked section;

```dart

_profileTopSection() {
    return SizedBox(
      height: 340, // This is the MAGIC trick
      child: Stack(
        alignment: Alignment.center,
        clipBehavior: Clip.none,
        children: [
          Positioned(
            top: 0, // This pushes the _heroSection imple to the topmost section of the screen, covering the AppBar(that's intended)
            child: _heroSection(),
          ),
          Positioned(
            top: heroSectionHeight - (_profileHeight / 1.4), // This is to just get a better placement for the profile image
            child: _profileImage(),
          ),
        ],
      ),
    );
  }





```

3. Now, with the height set to it, I get to reveal the stacked items and expose them to receive the hit signature.

https://github.com/user-attachments/assets/a5b71b22-33fd-4550-9324-220b32baf4d5
